src/compiler/program.ts

```diff
@@ -599,99 +636,106 @@ namespace ts {
+                        }
 
-                if (oldSourceFile !== newSourceFile) {
-                    if (oldSourceFile.hasNoDefaultLib !== newSourceFile.hasNoDefaultLib) {
-                        // value of no-default-lib has changed
-                        // this will affect if default library is injected into the list of files
-                        return false;
-                    }
+                        // check imports and module augmentations
+                        collectExternalModuleReferences(newSourceFile);
+                        if (!arrayIsEqualTo(oldSourceFile.imports, newSourceFile.imports, moduleNameIsEqualTo)) {
+                            // imports has changed
+                            structuralChanges |= StructuralChangesFromOldProgram.Imports;
+                        }
+                        if (!arrayIsEqualTo(oldSourceFile.moduleAugmentations, newSourceFile.moduleAugmentations, moduleNameIsEqualTo)) {
+                            // moduleAugmentations has changed
+                            structuralChanges |= StructuralChangesFromOldProgram.ModuleAugmentations;
+                        }
 
-                    // check tripleslash references
-                    if (!arrayIsEqualTo(oldSourceFile.referencedFiles, newSourceFile.referencedFiles, fileReferenceIsEqualTo)) {
-                        // tripleslash references has changed
-                        return false;
-                    }
+                        if (!arrayIsEqualTo(oldSourceFile.typeReferenceDirectives, newSourceFile.typeReferenceDirectives, fileReferenceIsEqualTo)) {
+                            // 'types' references has changed
+                            structuralChanges |= StructuralChangesFromOldProgram.TripleSlashTypesReferences;
+                        }
 
-                    // check imports and module augmentations
-                    collectExternalModuleReferences(newSourceFile);
-                    if (!arrayIsEqualTo(oldSourceFile.imports, newSourceFile.imports, moduleNameIsEqualTo)) {
-                        // imports has changed
-                        return false;
+                        // tentatively approve the file
+                        modifiedSourceFiles.push({ oldFile: oldSourceFile, newFile: newSourceFile });
                     }
-                    if (!arrayIsEqualTo(oldSourceFile.moduleAugmentations, newSourceFile.moduleAugmentations, moduleNameIsEqualTo)) {
-                        // moduleAugmentations has changed
-                        return false;
-                    }
-
-                    if (!arrayIsEqualTo(oldSourceFile.typeReferenceDirectives, newSourceFile.typeReferenceDirectives, fileReferenceIsEqualTo)) {
-                        // 'types' references has changed
-                        return false;
+                    else {
+                        // file has no changes - use it as is
+                        newSourceFile = oldSourceFile;

```

Comment:
It's a bit hard to tell in this diff... but is this in the else clause to `if (oldSourceFile !== newSourceFile)`? If so, this statement seems like a no-op.